roboRIO-2024/src/main/java/frc/robot/subsystems/Odometry.java

```diff
@@ -0,0 +1,132 @@
+import edu.wpi.first.math.kinematics.DifferentialDriveKinematics;
+import edu.wpi.first.math.kinematics.DifferentialDriveOdometry;
+import edu.wpi.first.math.kinematics.DifferentialDriveWheelSpeeds;
+import edu.wpi.first.units.BaseUnits;
+import edu.wpi.first.units.Unit;
+import edu.wpi.first.units.Units;
+import edu.wpi.first.wpilibj.DriverStation;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+import frc.robot.Settings;
+
+/**<h1>i hate this **/
+public class Odometry extends SubsystemBase{
+
+
+    Pose2d pose = new Pose2d();
+
+    PigeonIMU gyro;
+    RelativeEncoder leftEncoder;
+    RelativeEncoder rightEncoder;
+    Drivetrain drivetrain = new Drivetrain();
+    
+
+    PIDController leftController = new PIDController(0.00001, 0, 1);
+    PIDController rightController = new PIDController(0.00001, 0, 1);
+
+    private static final double kTrackWidth = 0.381 * 2; // meters, this is the defauklt from wpilib
+                                                         // change this later
+
+    private final DifferentialDriveKinematics kinematics =
+      new DifferentialDriveKinematics(kTrackWidth);
+
+    
+    private final SimpleMotorFeedforward feedforward = new SimpleMotorFeedforward(1, 3);
+    
+    // constructor so i can find in in the wall of code
+    public Odometry () {
+        gyro = drivetrain.getGyro();
+        leftEncoder = drivetrain.getLeftEncoder();
+        rightEncoder = drivetrain.getRightEncoder();
+
+        AutoBuilder.configureRamsete(
+                this::getPose, // Robot pose supplier
+                this::resetPose, // Method to reset odometry (will be called if your auto has a starting pose)
+                this::getCurrentSpeeds, // Current ChassisSpeeds supplier
+                this::driveChassisSpeeds, // Method that will drive the robot given ChassisSpeeds
+                new ReplanningConfig(), // Default path replanning config. See the API for the options here
+                () -> {
+                    // Boolean supplier that controls when the path will be mirrored for the red alliance
+                    // This will flip the path being followed to the red side of the field.
+                    // THE ORIGIN WILL REMAIN ON THE BLUE SIDE
+
+                    var alliance = DriverStation.getAlliance();
+                    if (alliance.isPresent()) {
+                        return alliance.get() == DriverStation.Alliance.Red;
+                    }
+                    return false;
+                },
+                this // Reference to this subsystem to set requirements
+        );
+    }
+
+    DifferentialDriveOdometry m_odometry = new DifferentialDriveOdometry(
+  new Rotation2d(BaseUnits.Angle.convertFrom(gyro.getYaw(), Units.Degrees)),
+  leftEncoder.getPosition(), rightEncoder.getPosition(),
+  new Pose2d(5.0, 13.5, new Rotation2d()));
+
+  @Override
+  public void periodic() {
+    var gyroAngle = gyro.getYaw();
```

Comment:
Probably need to convert this to radians. I would create a function getRotation() and that returns a Rotation2d do the conversion there, than call that when it's needed. This function should also make sure that the rotation is counterclockwise positive. I believe Will already wrote a function like this in Drivetrain.